ext/drx/drx.c

```diff
@@ -1372,3 +1405,1117 @@ drx_tail_pad_block(void *drcontext, instrlist_t *ilist)
+                    goto drx_expand_scatter_gather_exit;
+                if (!expand_avx512_scatter_gather_extract_scalar_index(
+                        drcontext, bb, sg_instr, el, &sg_info, scratch_xmm, scratch_reg0,
+                        orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                reg_id_t scalar_index_reg = scratch_reg0;
+                if (!expand_gather_load_scalar_value(drcontext, bb, sg_instr, &sg_info,
+                                                     scalar_index_reg, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                reg_id_t scalar_value_reg = scratch_reg0;
+                if (!expand_avx512_gather_insert_scalar_value(drcontext, bb, sg_instr, el,
+                                                              &sg_info, scalar_value_reg,
+                                                              scratch_xmm, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                if (!expand_avx512_scatter_gather_update_mask(
+                        drcontext, bb, sg_instr, el, &sg_info, scratch_reg0, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                MINSERT(bb, sg_instr, skip_label);
+            }
+        } else /* AVX-512 instr_is_scatter(sg_instr) */ {
+            for (uint el = 0; el < no_of_elements; ++el) {
+                instr_t *skip_label = INSTR_CREATE_label(drcontext);
+                expand_avx512_scatter_gather_make_test(drcontext, bb, sg_instr, el,
+                                                       &sg_info, scratch_reg0, skip_label,
+                                                       orig_app_pc);
+                if (!expand_avx512_scatter_gather_extract_scalar_index(
+                        drcontext, bb, sg_instr, el, &sg_info, scratch_xmm, scratch_reg0,
+                        orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                reg_id_t scalar_index_reg = scratch_reg0;
+                reg_id_t scalar_value_reg = scratch_reg1;
+                if (!expand_avx512_scatter_extract_scalar_value(
+                        drcontext, bb, sg_instr, el, &sg_info, scratch_xmm,
+                        scalar_value_reg, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                if (!expand_avx512_scatter_store_scalar_value(
+                        drcontext, bb, sg_instr, &sg_info, scalar_index_reg,
+                        scalar_value_reg, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                if (!expand_avx512_scatter_gather_update_mask(
+                        drcontext, bb, sg_instr, el, &sg_info, scratch_reg0, orig_app_pc))
+                    goto drx_expand_scatter_gather_exit;
+                MINSERT(bb, sg_instr, skip_label);
+            }
+        }
+        /* The mask register is zeroed completely when instruction finishes. */
```

Comment:
Wait: you're not zeroing any part of the mask until the very end?  That does not sound like it will work.  We need to be able to re-start the instruction in the middle, and isn't the mask the key piece of state to accomplish that?  I would expect it to have to be updated after each load/store.

Reply:
Read the description of the sequence. The mask _is_ updated after every load/store. But the instruction also zeroes the entire mask register, not only the element bits. The only right place to do this in the scalar sequence is at the end. And this will need handling in the restore code later.

Reply:
Right, I see.